src/test/regress/sql/segspace.sql

```diff
@@ -327,3 +327,25 @@ end;
 $func$ language plpgsql;
 
 select workset_cleanup_test();
+
+------------ Ensure that tuplestore is destroyed correctly in a case of error -------------------
+
+create table testdata as
+(select
+  md5(random()::text) as textdata,
+  random() * i as numdata,
+  timestamp '2000-01-01 00:00:00' + random() *
+    (timestamp '2024-01-01 00:00:00' - timestamp '2000-01-01 00:00:00') as datedata
+from generate_series(1, 1000000) i);
+
+set statement_mem = 200;
```

Comment:
```
postgres=# set statement_mem = 200;
ERROR:  200 is outside the valid range for parameter "statement_mem" (1000 .. 2147483647)
```
?

Reply:
Another question is do we need this GUC at all? It seems we don't need it (at least in test terms).

Reply:
This guc forces postgres to spill. If it won't spill we then the problem wouldn't reproduce

Reply:
Looks strange. I run the following:
```
create table testdata as
(select
  md5(random()::text) as textdata,
  random() * i as numdata,
  timestamp '2000-01-01 00:00:00' + random() *
    (timestamp '2024-01-01 00:00:00' - timestamp '2000-01-01 00:00:00') as datedata
from generate_series(1, 1000000) i);

--set statement_mem = 200;

set gp_workfile_limit_per_query = 100;

begin;
declare testdata_cursor cursor without hold for select * from testdata;
fetch forward all from testdata_cursor;
rollback;

reset statement_mem;
reset gp_workfile_limit_per_query;
drop table testdata;
```
And get the following stack trace during SIGSEGV
```
#0  __pthread_kill_implementation (no_tid=0, signo=11, threadid=136990798711168) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=11, threadid=136990798711168) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=136990798711168, signo=signo@entry=11) at ./nptl/pthread_kill.c:89
#3  0x00007c97a8442476 in __GI_raise (sig=11) at ../sysdeps/posix/raise.c:26
#4  0x00005ad2f3f7a89c in StandardHandlerForSigillSigsegvSigbus_OnMainThread (processName=0x5ad2f4758f83 "Master process", 
    postgres_signal_arg=11) at elog.c:5588
#5  0x00005ad2f3ddb454 in CdbProgramErrorHandler (postgres_signal_arg=11) at postgres.c:3597
#6  0x00005ad2f44cc454 in wrapper_handler (postgres_signal_arg=11) at pqsignal.c:98
#7  <signal handler called>
#8  pfree (pointer=0x5ad2fb621c20) at mcxt.c:1422
#9  0x00005ad2f3ff0cca in tuplestore_end (state=0x5ad2fb621998) at tuplestore.c:507
#10 0x00005ad2f3fc0840 in PortalDrop (portal=0x5ad2fb619918, isTopCommit=0 '\000') at portalmem.c:614
#11 0x00005ad2f3fbfe05 in CreatePortal (name=0x5ad2f475826c "", allowDup=1 '\001', dupSilent=1 '\001') at portalmem.c:220
#12 0x00005ad2f3dd8171 in exec_simple_query (query_string=0x5ad2fb6c8a18 "rollback;") at postgres.c:1809
#13 0x00005ad2f3dddf53 in PostgresMain (argc=1, argv=0x5ad2fb5d8b50, dbname=0x5ad2fb5d89b0 "postgres", 
    username=0x5ad2fb5d8988 "innerlife") at postgres.c:5379
```
Not sure, but looks what we fix like.

Reply:
Are you sure you're using patched version? Patch works for me both in tests and in a manual run

Reply:
I run it on not patched version to show that set statement_mem may be unnecessary.

Reply:
This may have worked on your host with a little amount of memory but I'm not sure that it will be the case if we run this test on a machine with significant amount of ram. This test hugely depends on memory limitations

Reply:
Looking to statement_mem I see it mainly used with resource queue/groups. Do we have an example of test (it may be a historical log from CI), which triggers an error because of low statement_mem?

Reply:
The problem on customer's cluster may have occurred because of enabled workfile compression which consumes a lot of memory. This should've exhausted memory and forced gpdb to spill. This error can be reproduced only when spilling happens, so the amount of ram is the main aspect. You can make sure by lowering the number of tuples in table from 1000000 to 1000, for example. You'll see that if postgres doesn't spill then no error occurs